general/chainerrl/baselines/agents/gail.py

```diff
@@ -0,0 +1,318 @@
+        action = self.xp.array(action, dtype=self.xp.float32)
+        return self.get_batch_reward(self.xp.expand_dims(obs, axis=0),
+                                     self.xp.expand_dims(action, axis=0))[0]
+
+    def get_reward_and_train(self, obs, action, expert_obs, expert_action):
+        obs = self.xp.array(obs, dtype=self.xp.float32)
+        action = self.xp.array(action, dtype=self.xp.float32)
+        expert_obs = self.xp.array(expert_obs, dtype=self.xp.float32)
+        expert_action = self.xp.array(expert_action, dtype=self.xp.float32)
+        reward = self.get_reward(obs, action)
+        self.trajectories_fake_obs.append(obs)
+        self.trajectories_fake_action.append(action)
+        self.trajectories_true_obs.append(expert_obs)
+        self.trajectories_true_action.append(expert_action)
+        self._update_if_dataset_is_ready()
+        return reward
+
+    def get_batch_reward(self, batch_obs, batch_action):
+        batch_obs = self.xp.array(batch_obs, dtype=self.xp.float32)
+        batch_action = self.xp.array(batch_action, dtype=self.xp.float32)
+        if self.obs_normalizer is not None:
+            batch_obs = self.obs_normalizer(batch_obs, update=False)
+        with chainer.using_config('train', False), chainer.no_backprop_mode():
+            infer = self.model(batch_obs, batch_action)
+            return chainer.cuda.to_cpu(
+                -F.log(1 - F.sigmoid(infer) + 1e-8).array)[:, 0]
+
+    def get_batch_reward_and_train(self, batch_obs, batch_action,
+                                   batch_expert_obs, batch_expert_action):
+        batch_obs = self.xp.array(batch_obs, dtype=self.xp.float32)
+        batch_action = self.xp.array(batch_action, dtype=self.xp.float32)
+        batch_reward = self.get_batch_reward(batch_obs, batch_action)
+        self.trajectories_fake_obs += list(batch_obs)
+        self.trajectories_fake_action += list(batch_action)
+        self.trajectories_true_obs += list(batch_expert_obs)
+        self.trajectories_true_action += list(batch_expert_action)
+        self._update_if_dataset_is_ready()
+        return batch_reward
+
+    def _update_if_dataset_is_ready(self):
+        if len(self.trajectories_fake_obs) >= self.update_interval:
+            self._update()
+            self._reset_trajectories()
+
+    def _loss(self, fake_batch_obs, fake_batch_action,
+              true_batch_obs, true_batch_action):
+        if self.obs_normalizer is not None:
+            normalized_obs = self.obs_normalizer(fake_batch_obs, update=False)
+            infer_fake = self.model(normalized_obs, fake_batch_action)
+        else:
+            infer_fake = self.model(fake_batch_obs, fake_batch_action)
+        if self.noisy_label:
+            n = fake_batch_obs.shape[0]
+            fake_loss = -F.average(
+                F.log(F.absolute(1 - self.xp.random.rand(n) * 0.3
+                                 - F.sigmoid(infer_fake))
+                      + 1e-8))
```

Comment:
maybe adding arguments for the 1e-8 and 0.3 constants would be better to ensure they're consistent in case some wants to change them